### ISSUE TITLE
Attempting to select maximum values does not work as expected

### ISSUE THREAD

**Issue by mmcgr** [souffle-lang/souffle#550]
I would expect this code to find b(100,"z"):
```
.decl a(x:number, y:symbol)
a(1,"x").
a(10,"y").
a(100, "z").

.decl b(x:number, y:symbol)
b(x, y) :- x = max z : { a(z,y) }, a(x, y).
//b(max  z0 : a( z0,_),y) :-  a(max  z0 : a( z0,_),y).

.output b()
```
It actually finds all the values in a.
```
---------------
b
x	y
===============
1	x
10	y
100	z
===============
```
The debug report says the rule for `b` is transformed into the commented out version. However, if you run that the result is actually
```
---------------
b
x	y
===============
100	x
100	y
100	z
===============
```

**Comment by azreika** [comment]
When I run it locally, the debug report actually shows the rule:
`b(x, y) :- x = max z : { a(z,y) }, a(x, y).`
being transformed into
`b(max  z0 : a( z0,y),y) :- a(max  z0 : a( z0,y),y).`
which is different to the one you had commented out because the `y`'s inside the aggregators here are still tied to the external `y`'s (as opposed to just being free variables). If you run that, it should produce the same result as your first example.

Either way, I think the output for your first result is expected. Since witnesses aren't passed to the outer scope, the variable has to be grounded in the outer scope first. This means we look at `a(x,y)` first, which grounds the values of `x` and `y`, and affects the max aggregator.
We can check if `b(1,"x")` is true by just plugging in those values into the clause directly. In this case, the RHS becomes:
`1 = max z : { a(z,"x") }, a(1,"x")`
But `max z : { a(z,"x") }` evaluates to 1 because `a(1,"x")` is the only related tuple, and so we end up with
`1 = 1, a(1,"x")` 
which is true, and so `b(1,"x")` must be true.

That doesn't mean the semantics aren't broken, though, because the following example produces the wrong output:

```
.decl a(x:number, y:symbol)
a(1,"x").
a(100,"x").

a(10,"y").
a(100,"y").

.decl b(x:number, y:symbol)
b(x, y) :- x = max z : { a(z,y) }, a(x, y).

.output b()
```

This outputs

```
---------------
b
x	y
===============
1	x
100	x
1	y
100	y
===============
```

But if we try plugging in `(x,y) = (1,"x")` into the RHS now we get:
`1 = max z : { a(z,"x") }, a(1,"x")`
This time, `max z : { a(z,"x") }` should evaluate to 100, which is not equal to 1, so this should not be true - meaning the output is wrong.

In the debug report, though, the clause is transformed to `b(max  z0 : a( z0,y),y) :- a(max  z0 : a( z0,y),y).`, and if you use that directly in the original program then you get the expected result of:

```
---------------
b
x	y
===============
100	x
100	y
===============
```

If you look at the debug report of *this* newly changed program then `b(max  z0 : a( z0,y),y) :- a(max  z0 : a( z0,y),y).` is transformed to `b(max  z00 : a( z00,y),y) :- a(max  z01 : a( z01,y),y).`, so it looks like it's possibly an issue with variable naming/scope.

**Comment by b-scholz** [comment]
As discussed the last few days - this needs to be fixed. Perhaps we look into this next week. There is another aggregate. See issue #556.

**Comment by b-scholz** [comment]
We should stop using the transformer for more complex aggregates that constructs new auxiliary tables. Instead, we should build a local loop-nest for each aggregate inside a loop-nest.

I will discuss this with Martin.

**Comment by azreika** [comment]
Seems like it could be a naming issue related to the order of the AST transformations.

Right now, the `UniqueAggregationVariablesTransformer` transformation is applied before the `ResolveAliasesTransformer` transformation. The former makes unique all variables we are aggregating over, while the latter might introduce new aggregators by resolving equalities.

E.g. in the program above we have the clause:

`b(x, y) :- x = max z : { a(z,y) }, a(x, y).`

The `UniqueAggregationVariablesTransformer` transformation changes this first to:

`b(x,y) :- x = max z_0 : { a(z_0,y) }, a(x,y).`

And then the `ResolveAliasesTransformer` is run to change this into:

`b(max z_0 : { a(z_0,y) },y) :- a(max z_0 : { a(z_0,y) },y).`

In the execution of the program, the output will be the incorrect one:

```
---------------
b
x	y
===============
1	x
100	x
1	y
100	y
===============
```

Notice how there are non-uniquely-named aggregator variables, namely `z_0`, in the same clause.

If we again apply the `UniqueAggregationVariablesTransformer` right after the `ResolveAliasesTransformer`, though, then we get:

`b(max z_0_0 : { a(z_0_0,y) },y) :- a(max z_0_1 : { a(z_0_1,y) },y).`

and so we now _actually_ have unique aggregator variables.

In the execution of the program, the output will be the correct one:

```
---------------
b
x	y
===============
100	x
100	y
===============
```
So re-applying the `UniqueAggregationVariablesTransformer` after resolving aliases seems to solve the issue.

RAM without the re-application of `UniqueAggregationVariablesTransformer` after resolving aliases:
```
	INSERT 
		SCAN a AS t0 
			MAX t1.x AS t1.0 IN t1 ∈ a(t0.x,t0.y) FOR 
				PROJECT (env(t1, i0), t0.y) INTO b
```

RAM with the re-application:
```
        INSERT 
		SCAN a AS t0 
			MAX t1.x AS t1.0 IN t1 ∈ a(_,t0.y) FOR 
				MAX t2.x AS t2.0 IN t2 ∈ a(t0.x,t0.y) FOR 
					PROJECT (env(t1, i0), t0.y) INTO b
```

Not entirely sure why they differ, though, even when not uniquely named, because the aggregator variables should still be aggregating over identical bodies.

**Comment by mmcgr** [comment]
Closed in favour of #662.
